scripts/irods/controller.py

```diff
@@ -155,7 +155,7 @@ def reload_configuration(self):
             if (new_agent_factory_pid is None) or (new_agent_factory_pid == old_agent_factory_pid):
                 time.sleep(1)
                 loop_counter += 1
-                if loop_counter == 15:
+                if loop_counter == 60:
```

Comment:
interesting.    that seems... high.    it had to sleep for more than 15 seconds?

any thoughts on why?

Reply:
I agree 60 is a long time. Raising that value was due to the number of containers running simultaneously, so this is temporary.

However, while the value is questionable, an admin is unlikely to ever hit that because they aren't running as many servers as us on a single computer.

We need to discuss whether this is the right approach. I'll list what I've done prior to this.
- 1st impl: Wait until the agent factory returned a successful heartbeat response
  - This was bad because it didn't consider the possibility of config validation failing and there was no way to ask the server what happened due to the use of signals
- 2nd impl: Wait until the agent factory PID changed and test using a heartbeat
  - This is better, but suffers the same issue as the original impl
- 3rd impl: This is the current impl, which is the same as the 2nd impl + a loop counter
  - This approach eventually returns control to the admin and instructs them to check the server

Reply:
I also want to mention that the bump to 60 is a strong vote that the value should be configurable, even if only for testing.

One could argue that needing a counter at all means we should rethink the approach, hmm.

Reply:
can we check the config on the new one BEFORE we send a signal to the current agent factory?

Reply:
> the value should be configurable

perhaps.   but then again... a minute is a long time... means something is really wrong, and it's just a backstop...

a knob to allow the backstop to be 30 seconds doesn't really change anything about the code or the deployment itself that would actually need to be addressed.

Reply:
> can we check the config on the new one BEFORE we send a signal to the current agent factory?

That's a possibility. I decided against that because it meant validating the config twice. Once in python and a second time by the server when it receives the SIGHUP.

I left out that there was another option. That is, send the signal and don't wait by default. The python function would expose a parameter which allows testing to force the python code to wait until the reload completes.

When an admin invokes reload, it would be equivalent to calling `kill -HUP` and that's it. Then there's no possibility for errors or failure python-wise.

Reply:
> it meant validating the config twice. Once in python and a second time by the server when it receives the SIGHUP.

this is not a performance thing - so… is it so bad to check it twice if it removes the ‘broken’ case in 1st impl?

Reply:
True, but doing the same work twice feels like something isn't designed quite right.

With the current implementation of iRODS 5, there's no real reason to keep `irodsctl` since its primary function can be accomplished directly through signals. Not only that, but use of systemd will be the preferred way of controlling iRODS 5 when it becomes available.

The `irodsctl upgrade` command can be exposed through a separate python script.

Thoughts?

Reply:
I forgot to mention that the functionality provided by controller.py is effectively for setup_irods.py and the test suite now. Nothing else, outside of irodsctl, relies on the logic in question.

Reply:
> The irodsctl upgrade command can be exposed through a separate python script. 

i think that makes sense... an `upgrade_irods.py` script would have one job.

and...

your fourth option of having a python 'test' option to 'wait' is interesting...
 ... but does that account for a bad config?  it would still wait forever if there was a bad config...
  still feels like the cost of having signal-only communication and no return information means we should check BEFORE sending the signal... if we have to assume the agent factory will come back up.

Reply:
For testing, I don't think we have to be concerned about bad configs because the tests control everything. A bad config only happens if the developer makes a mistake or intentionally designed it that way.

We should look at what nginx/httpd do when reloading with a bad config.
- Does systemd wait for the reload to complete or return immediately, regardless of the outcome?
- Does systemd or the application monitored by systemd report anything when a bad config is present? Is the error message relayed by systemd?

I think answering these questions could help us.

Reply:
After discussing with the team, we've agreed to do the following:
- Remove the `irodsctl` script
  - The current server implementation can be used with systemd quite easily
- We can introduce an option for instructing the python reload function to increase the loop counter value
  - This would be aimed at testing for the most part
  - The option can either be defined as part of the reload function signature or picked up via an environment variable

Reply:
Turns out the testing environment relies on irodsctl to start/stop servers. If we remove irodsctl, then we'd need to come up with a way for the testing environment to achieve the same behavior without irodsctl.

I'm going to leave irodsctl in for now and address its removal in a later PR.

Reply:
Open an issue for considering the removal of irodsctl.

Reply:
Created #7996

Reply:
I've opened a PR for the testing environment which replaces use of `irodsctl` with the `IrodsController.`

See irods/irods_testing_environment#231 for details.

I'll be removing `irodsctl` as part this PR.

Reply:
I've confirmed that the removal of `irodsctl` in 41e108b work.

When a server is running and an upgrade occurs, the packaging will stop the server by calling `IrodsController().stop()` and then proceed with the upgrade.

The removal of `irodsctl` means admins must run `irodsServer -d` to start the server as a daemon now. This is intentional.

A systemd service file will be provided in place of `irodsctl`.

Reply:
got it - makes sense.